src/connect.rs

```diff
@@ -0,0 +1,47 @@
+use std::fmt;
+use std::io::Result as IoResult;
+use std::net::{SocketAddr, TcpStream};
+use std::sync::Arc;
+use std::time::Duration;
+
+/// A custom Connector to override the default TcpStream connector.
+pub trait Connector: Send + Sync {
+    fn connect(&self, addr: &SocketAddr) -> IoResult<TcpStream> {
+        TcpStream::connect(addr)
+    }
+
+    fn connect_timeout(&self, addr: &SocketAddr, timeout: Duration) -> IoResult<TcpStream> {
+        TcpStream::connect_timeout(addr, timeout)
+    }
+}
```

Comment:
I think this should tie into the `ReadWrite` trait. https://github.com/algesten/ureq/blob/4bb6d3a4dbd7acafed801dae9d75df02207e1b87/src/stream.rs#L21

I.e. the connector trait should not produce a concrete `std::net::TcpStream`, but rather something that implements `ReadWrite`. I haven't followed through all the trait bound implications, but the starting point would be:

```rs
pub trait Connector: Send + fmt::Debug + 'static {
    type Socket;

    fn connect(&self, addr: &SocketAddr) -> io::Result<Self::Socket>;
    fn connect_timeout(&self, addr: &SocketAddr, timeout: Duration) -> io::Result<Self::Socket>;
    fn set_read_timeout(&self, socket: &mut Self::Socket);
}
```

and the agent would have something like:

```rs
    pub connector: Arc<dyn Connector<Socket = dyn ReadWrite>>,
```

Notice this abstraction would require use redo today's handling of `set_read_timeout` relying on `ReadWrite::socket()`. Unclear whether we should just go `set_read_timeout` straight on the `ReadWrite` trait.

Also minor detail, style wise we use `io::Error` instead of `IoError`.

Reply:
`ureq` relies on the functions of `TcpStream` everywhere.
If we want to abstract it into a trait, we need to cover almost all of `TcpSteam`'s methods, I think this introduces too much complexity since `TcpStream` is already an abstraction for multiple platforms.

It looks like `hyper` adds a layer to be compatible with tcp-based h2 and quic-based h3, I don't think there is such a trait that can handle subsequent changes. I think abstracting away `TcpStream` should not be considered if not for the time being.

Reply:
> If we want to abstract it into a trait, we need to cover almost all of `TcpSteam`'s methods

Really? In my mind it's pretty much only `set_read_timeout` – but maybe I'm not thinking this through?

Reply:
and `set_write_timeout`, `set_nonblocking`, `set_nodelay`

Reply:
* `set_nodelay` can be set as soon as the socket has connected, no need to abstract over it (pass it as an argument to connect).
* `set_write_timeout` and `set_read_timeout` can probably be one single thing.
* `set_nonblocking` is used in a very specific way that can be abstracted.

I think we'd end up with something like:

```rs
pub trait Connector: Send + fmt::Debug + 'static {
    type Socket;

    fn connect(&self, addr: &SocketAddr, timeout: Option<Duration>, nodelay: bool) -> io::Result<Self::Socket>;
    fn set_read_write_timeout(&self, socket: &mut Self::Socket) -> io::Result<()>;
    fn check_server_closed(&self, socket: &mut Self::Socket) -> io::Result<()>;
}
```

Sure. It's a bigger change working this into the entire code base. Before embarking on it, maybe we should broaden the discussion.

@jsha what are your thoughts on this?

Reply:
Agree, this requires more discussion.